pallets/interest-accrual/src/lib.rs

```diff
@@ -437,118 +436,137 @@ pub mod pallet {
+									reference_count: 1,
+								};
+
+								rates
+									.try_push(new_rate)
+									.map_err(|_| Error::<T>::TooManyRates)?;
+
+								Ok(())
+							}
+						}
+					})
 				}
-			})
+			}
 		}
 
 		pub fn unreference_interest_rate(
-			interest_rate_per_year: T::InterestRate,
+			interest_rate_per_year: InterestRate<T::Rate>,
 		) -> DispatchResult {
-			let interest_rate_per_sec = unchecked_conversion(interest_rate_per_year)?;
-			Rates::<T>::try_mutate(|rates| {
-				let idx = rates
-					.iter()
-					.enumerate()
-					.find(|(_, rate)| rate.interest_rate_per_sec == interest_rate_per_sec)
-					.ok_or(Error::<T>::NoSuchRate)?
-					.0;
-				rates[idx].reference_count = rates[idx].reference_count.saturating_sub(1);
-				if rates[idx].reference_count == 0 {
-					rates.swap_remove(idx);
+			match interest_rate_per_year {
+				InterestRate::Fixed { rate_per_year, .. } => {
+					let interest_rate_per_sec = unchecked_conversion(rate_per_year)?;
+					Rates::<T>::try_mutate(|rates| {
+						let idx = rates
+							.iter()
+							.enumerate()
+							.find(|(_, rate)| rate.interest_rate_per_sec == interest_rate_per_sec)
+							.ok_or(Error::<T>::NoSuchRate)?
+							.0;
+						rates[idx].reference_count = rates[idx].reference_count.saturating_sub(1);
+						if rates[idx].reference_count == 0 {
+							rates.swap_remove(idx);
+						}
+						Ok(())
+					})
 				}
-				Ok(())
-			})
+			}
 		}
 
 		pub fn get_rate(
-			interest_rate_per_year: T::InterestRate,
+			interest_rate_per_year: InterestRate<T::Rate>,
 		) -> Result<RateDetailsOf<T>, DispatchError> {
-			let interest_rate_per_sec = unchecked_conversion(interest_rate_per_year)?;
-			Rates::<T>::get()
-				.into_iter()
-				.find(|rate| rate.interest_rate_per_sec == interest_rate_per_sec)
-				.ok_or_else(|| Error::<T>::NoSuchRate.into())
+			match interest_rate_per_year {
+				InterestRate::Fixed { rate_per_year, .. } => {
+					let interest_rate_per_sec = unchecked_conversion(rate_per_year)?;
+					Rates::<T>::get()
+						.into_iter()
+						.find(|rate| rate.interest_rate_per_sec == interest_rate_per_sec)
+						.ok_or_else(|| Error::<T>::NoSuchRate.into())
+				}
+			}
 		}
 
 		pub(crate) fn validate_interest_rate(
-			interest_rate_per_year: T::InterestRate,
+			interest_rate_per_year: InterestRate<T::Rate>,
 		) -> DispatchResult {
-			let four_decimals = T::InterestRate::saturating_from_integer(10000);
-			let maximum = T::InterestRate::saturating_from_integer(MAX_INTEREST_RATE);
-			ensure!(
-				interest_rate_per_year <= maximum
-					&& interest_rate_per_year >= Zero::zero()
-					&& (interest_rate_per_year.saturating_mul(four_decimals)).frac()
-						== Zero::zero(),
-				Error::<T>::InvalidRate
-			);
-			Ok(())
+			match interest_rate_per_year {
+				InterestRate::Fixed { rate_per_year, .. } => {
+					let four_decimals = T::Rate::saturating_from_integer(10000);
+					let maximum = T::Rate::saturating_from_integer(MAX_INTEREST_RATE);
```

Comment:
Rather for @branan and @lemunozm. Why are we using saturating math here?

Reply:
Because `MAX_INTEREST_RATE` is a const, we can ensure it's ok at compile time. But we could use `check_from_integer()` too, to be sure the constant value is not changed after the conversion